src/mds/Server.cc

```diff
@@ -1242,6 +1251,15 @@ void Server::evict_cap_revoke_non_responders() {
   }
 
   auto&& to_evict = mds->locker->get_late_revoking_clients(cap_revoke_eviction_timeout);
+  // don't evict client(s) if osds are laggy
+  bool laggy = mds->objecter->with_osdmap([](const OSDMap &map) {
+    return map.any_osd_laggy(); });
+  if(to_evict.size() && laggy) {
+    mds->get_beacon().notify_health(mds);
+    dout(10) << "laggy OSDs deteced, skipping laggy clients eviction " 
+             << "untill they return to normal" << dendl;
+    return;
```

Comment:
So the Beacon code does not expect to have `notify_health` called so frequently. It should be called only from `MDSRank::tick` (or in disaster cases). Also, Beacon is written to call into other modules once it has the `mds_lock`, not the other way around.

So I would recommend:

- create a `std::set<Session*> Server::laggy_sessions` member.
- Clear the set in `MDSRank::tick` before calling the related Server methods
- Have beacon call a `Server::` helper method to get the size of the set to put in the beacon message.

Reply:
two questions:
1) Didn't quite understand the second point of clearing the set in `MDSRankDispatcher::tick`
2) Considering that Beacon should be the one calling Server and not the other way around, how to communicate the message(Server to Beacon) that we have some laggy sessions? have something like `have_laggy_sessions()` in `tick()` that calls in a helper methods from Server? if yes then could there be some sync issue between Server and `tick()` here?

Reply:
>Didn't quite understand the second point of clearing the set in MDSRankDispatcher::tick

So the set gives a point-in-time representation of laggy sessions. Otherwise, you have to do work to remove sessions which are no longer laggy.

> Considering that Beacon should be the one calling Server and not the other way around, how to communicate the message(Server to Beacon) that we have some laggy sessions? have something like have_laggy_sessions() in tick() that calls in a helper methods from Server? if yes then could there be some sync issue between Server and tick() here?

Server would have a helper method that just returns the size of the `std::set`. (I think a better name now is `Server::sessions_laggy_osd`.) Have Beacon call that from `::notify_health`.

Reply:
ah, my bad, didnt check the tick() code before asking. i had the misconception that `find_idle_sessions` might be getting called somewhere else.

Reply:
instead of creating a set and recording all the idle/non-responding sessions, maybe it would be better to maintain a bool value `any_osd_laggy` and reset it before calling `find_idle_sessions` and `evict_cap_revoke_non_responders` in `tick()`

Reply:
we would miss on a lot of data as per above comment, therefore have implemented your suggestions. thanks!